### DISCUSSION TITLE
Nimbus support

### DISCUSSION THREAD

**Discussion by jsamol** [smol-dot/smoldot#1574] [All]
As far as I can tell, only Aura and BABE consensus models are supported in Smoldot. Are there maybe any plans to add support for [Nimbus](https://github.com/moonbeam-foundation/nimbus) as well?

**Comment by tomaka** [comment]
I looked at Nimbus around a year ago.
From what I gathered back then, since according to Nimbus a block is valid only after it has executed some runtime code, and that a light client doesn't execute runtime code to verify blocks, adding support for Nimbus would make the light client trivial to DoS attack.

**Reply by gitsimon** [reply]
@tomaka Could you elaborate, please? With Aura, the set of authorities can also change over time, by means of a dedicated RPC. Nimbus does calculate the set of authorities inside the runtime, but could expose this set via a RPC, called for the lastest known finalised block. Would that allow the generalisation of the block authoring mechanism?

Furthermore, if smoldot would open up for trivial DoS attacks, why does a Node running Nimbus-based authoring mechanism not?

**Reply by tomaka** [reply]
> With Aura, the set of authorities can also change over time, by means of a dedicated RPC.

What do you mean by "dedicated RPC"?

> Nimbus does calculate the set of authorities inside the runtime, but could expose this set via a RPC

The entire reason why smoldot and light clients exist is that they don't trust data blindly and need formal proofs. If it's acceptable to download the list of authorities via a JSON-RPC function, then you don't need a light client.

> if smoldot would open up for trivial DoS attacks, why does a Node running Nimbus-based authoring mechanism not?

There are two levels of verification for any given block: "authentic" and "valid". Making sure that a block is "authentic" consists in making sure that it has been authored by someone who had the rights to author it, and can be done just by reading the header. In order to make sure that a block is "valid", you need to execute the runtime code. A light client doesn't do that, as that would be too expensive, and stops at the "authentic" phase.

Making sure that blocks are authentic is necessary in order to avoid being spammed by bad blocks.
Unfortunately, with Nimbus, if I'm not mistaken, anyone can create an authentic block, and thus can spam the client until it runs out of memory or is forced to throw away legitimate blocks.

**Reply by gitsimon** [reply]
I meant to write "dedicated pallet function", more precisely [`change_authorities`](https://github.com/paritytech/polkadot-sdk/blob/e757f81110a92d200d72d419c67d945f46aa4489/substrate/frame/aura/src/lib.rs#L203), which is not directly exposed by an extrinsic or RPC.

Thanks for clarifying that every RPC call violates the verifiability of a block producer's authenticity.

Having a closer look at your code I figured out what you actually require from Aura, or the pallet aura: It seems to be **only** these [extra header fields](https://github.com/paritytech/polkadot-sdk/blob/e757f81110a92d200d72d419c67d945f46aa4489/substrate/primitives/consensus/aura/src/lib.rs#L78) that you then [parse](https://github.com/smol-dot/smoldot/blob/aa0eb433206c72b22f95d3f8fdd5bc171ea21e12/lib/src/header.rs#L1394) into this equally shaped [enum variant](https://github.com/smol-dot/smoldot/blob/0b7287e050146314b50252ccb657f363b54a8bbf/lib/src/header/aura.rs#L28).

Could you confirm that it is indifferent to smoldot **which pallet / consensus mechanism writes this header fields**, as long as
- it has the same encoding
- the new authorities are announced in a valid, final block's header _before_ they attempt to produce blocks
- can this header fields be added only at the first block of an epoch or at any block? From [this line](https://github.com/smol-dot/smoldot/blob/922abb97d9e973b73f5f12b4482d735e7d01f51e/lib/src/verify/aura.rs#L167) it seems it could be at any block

**Reply by tomaka** [reply]
> it is indifferent to smoldot which pallet / consensus mechanism writes this header fields

Yes, smoldot or clients in general don't know what a pallet is.

> the new authorities are announced in a valid, final block's header before they attempt to produce blocks

Aura is completely orthogonal to finalization. When new authorities are announced in a block, one of these authorities can author a child of that block.

This is what makes spam attacks powerful: a single malicious validator can author a fake block with a fake list of authorities whose keys it has generated locally, and then immediately author an infinitely-size chain.

> can this header fields be added only at the first block of an epoch or at any block? From [this line](https://github.com/smol-dot/smoldot/blob/922abb97d9e973b73f5f12b4482d735e7d01f51e/lib/src/verify/aura.rs#L167) it seems it could be at any block

There's no such thing as an "epoch" from the client's point of view. The client simply updates the list of of authorities whenever it finds this field, it doesn't care about anything else.
